### PR TITLE
[libc++][doc] Use installed std modules in external projects.

### PR DESCRIPTION
**Pull request by a858438680**
The original libc++ document about using std modules in external projects is outdated. Loot at #80231 for more details.

This pull request update the document. The new document tells how to use the current installed std modules without toolchains' support.

### CODE REVIEW

File: libcxx/docs/Modules.rst [hunk 1]
```diff
@@ -105,36 +105,36 @@ Users need to be able to build their own BMI files.
    system vendors, with the goal that building the BMI files is done by
    the build system.
 
-Currently this requires a local build of libc++ with modules enabled. Since
-modules are not part of the installation yet, they are used from the build
-directory. First libc++ needs to be build with module support enabled.
+Currently this requires a local build of libc++ with modules installation enabled.
+Since modules are not installed by default. You can build and install the modules
+to ``<install_prefix>`` with the following commands.
```
Comment (ChuanqiXu9) [inline libcxx/docs/Modules.rst:110]:
```suggestion
This requires the system you're distributes the installed std modules. Otherwise...
```

Comment (a858438680) [reply libcxx/docs/Modules.rst:110]: Thanks for your review. But I do not understand why you want me to delete these three lines. Could you please explain more in detail?
And also, The line you suggest me to add is not complete and seems to have a grammer error. Could you please tell me what you want to express?

Comment (ChuanqiXu9) [reply libcxx/docs/Modules.rst:110]: Sorry. I mean it may be better to not say "the modules are not installed by default". Since there may be system vendors enable it.

Comment (a858438680) [reply libcxx/docs/Modules.rst:110]: Ok, got it. So you mean "requires the system you're using distributes ..." right? And "Otherwise, you need to build and install ..." right? I think this suggestion is fine. But I also want to hear from main author of the original page @mordante.

File: libcxx/docs/Modules.rst [hunk 2]
```diff
@@ -168,58 +168,95 @@ This is a small sample program that uses the module ``std``. It consists of a
...
   #
 
-  add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-fprebuilt-module-path=${std_BINARY_DIR}/CMakeFiles/std.dir/>)
-  add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-fprebuilt-module-path=${std_BINARY_DIR}/CMakeFiles/std.compat.dir/>)
-  add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-nostdinc++>)
-  # The include path needs to be set to be able to use macros from headers.
-  # For example from, the headers <cassert> and <version>.
-  add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-isystem>)
-  add_compile_options($<$<COMPILE_LANGUAGE:CXX>:${LIBCXX_BUILD}/include/c++/v1>)
+  add_library(std)
```
Comment (poyaoc97) [inline libcxx/docs/Modules.rst:194]: Missing "STATIC"?

File: libcxx/docs/Modules.rst [hunk 3]
```diff
@@ -105,36 +105,36 @@ Users need to be able to build their own BMI files.
...
-modules are not part of the installation yet, they are used from the build
-directory. First libc++ needs to be build with module support enabled.
+Currently this requires a local build of libc++ with modules installation enabled.
+Since modules are not installed by default. You can build and install the modules
+to ``<install_prefix>`` with the following commands.
 
 .. code-block:: bash
 
-  $ git clone https://github.com/llvm/llvm-project.git
+  $ git clone https://github.com/llvm/llvm-project.git --depth 1
```
Comment (mordante) [inline libcxx/docs/Modules.rst:114]: I prefer to keep the instructions generic, if users like `--depth 1` they are free to do so.

File: libcxx/docs/Modules.rst (hunk 1 above)
Comment (mordante) [reply libcxx/docs/Modules.rst:110]: We shouldn't change this in this fashion. The current described way for building is broken, this should be fixed. So that way remains a viable option. The goals of this patch is to add a second option based on installed modules.

File: libcxx/docs/Modules.rst [hunk 4]
```diff
@@ -105,36 +105,36 @@ Users need to be able to build their own BMI files.
...
 
 .. code-block:: cpp
 
+  // main.cpp
   import std; // When importing std.compat it's not needed to import std.
   import std.compat;
 
   int main() {
-    std::cout << "Hello modular world\n";
+    std::println("Hello modular world");
```
Comment (mordante) [inline libcxx/docs/Modules.rst:131]: please don't make unrelated changes in the same patch.

**Review by mordante — CHANGES_REQUESTED** [review]
Thanks for the patch. I really like to keep the existing method and add a new method. I'll make a patch to fix to reintroduce the accidentally removed CMakeLists.txt.

**Comment by mordante** [comment]
@a858438680 are you still working on this patch?

**Comment by a858438680** [comment]
> @a858438680 are you still working on this patch?

Yes, but I missed your previous comments. I will make some changes according to your comments soon.

File: libcxx/docs/Modules.rst (hunk 3 above)
Comment (a858438680) [reply libcxx/docs/Modules.rst:114]: OK

File: libcxx/docs/Modules.rst (hunk 2 above)
Comment (a858438680) [reply libcxx/docs/Modules.rst:194]: I will add it

File: libcxx/docs/Modules.rst (hunk 4 above)
Comment (a858438680) [reply libcxx/docs/Modules.rst:131]: OK, will change it back

**Comment by a858438680** [comment]
@mordante Hi, I have changed this commit according the your comments.

**Comment by mordante** [comment]
> @mordante Hi, I have changed this commit according the your comments.

Thanks! Can you rebase the patch, there are merge conflicts. I prefer to review without these conflicts.

**Comment by a858438680** [comment]
@mordante Hi, there is no conflict now.

File: libcxx/docs/Modules.rst [hunk 5]
```diff
@@ -107,34 +107,40 @@ Users need to be able to build their own BMI files.
...
 
-  $ git clone https://github.com/llvm/llvm-project.git
+  $ git clone https://github.com/llvm/llvm-project.git --depth 1
   $ cd llvm-project
   $ mkdir build
-  $ cmake -G Ninja -S runtimes -B build -DLLVM_ENABLE_RUNTIMES="libcxx;libcxxabi;libunwind"
-  $ ninja -C build
+  $ cmake -G Ninja -S runtimes -B build -DLIBCXX_INSTALL_MODULES=ON -DLLVM_ENABLE_RUNTIMES="libcxx;libcxxabi;libunwind"
+  $ cmake --build build -- -j $(nproc)
+  $ cmake --install build --prefix <install_prefix>
```
Comment (mordante) [inline libcxx/docs/Modules.rst:122]: I really prefer to keep the `ninja -C build` and add `ninja install -C build`. This is the style we use in other parts of the libc++ documentation.

File: libcxx/docs/Modules.rst [hunk 6]
```diff
@@ -168,51 +174,82 @@ This is a small sample program that uses the module ``std``. It consists of a
...
+
+.. code-block:: cmake
 
+  # std.cmake
   include(FetchContent)
   FetchContent_Declare(
-    std
-    URL "file://${LIBCXX_BUILD}/modules/c++/v1/"
+    std_module
+    URL "file://${LIBCXX_INSTALLED_DIR}/share/libc++/v1"
```
Comment (mordante) [inline libcxx/docs/Modules.rst:187]: This part of the example should use the `${LIBCXX_BUILD}` right? Are the other parts of the changes in this part intended?

File: libcxx/docs/Modules.rst [hunk 7]
```diff
@@ -225,6 +231,137 @@ Building this project is done with the following steps, assuming the files
 .. note:: The ``std`` dependencies of ``std.compat`` is not always resolved when
           building the ``std`` target using multiple jobs.
 
+This is another small sample program that uses the module ``std`` from
+installation directory. It consists of a ``CMakeLists.txt``, an
+``std.cmake``, and a ``main.cpp`` file. The ``main.cpp`` is the same as
+the previous example.
```
Comment (mordante) [inline libcxx/docs/Modules.rst:237]:
```suggestion
the installation directory. It consists of a ``CMakeLists.txt``,
a ``std.cmake``, and the ``main.cpp`` from the previous example.
```

File: libcxx/docs/Modules.rst [hunk 8]
```diff
@@ -225,6 +231,137 @@ Building this project is done with the following steps, assuming the files
...
+  $ cmake -S . -B build -G Ninja -DCMAKE_CXX_COMPILER=<path-to-compiler> -DLIBCXX_INSTALLED_DIR=<install_prefix>
+  $ ninja -C build
+  $ build/main
+
+.. warning:: You need more than clang itself to build a project using modules.
+             Specifically, you will need ``clang-scan-deps``. For example, in Ubuntu, you
+             need to use ``sudo ./llvm.sh 17 all`` rather than ``sudo ./llvm.sh 17`` showed
+             in `LLVM Debian/Ubuntu nightly packages <https://apt.llvm.org>`__ to install
+             essential components to build this project.
+
```
Comment (mordante) [inline libcxx/docs/Modules.rst:364]: I would remove this part. This is also required when using the build directory.

**Review by mordante — CHANGES_REQUESTED** [review]
Thanks I finally had time to have a good look. I tested in locally and the instructions work nicely :-)

There have been some other changes in this file, can you rebase before updating this review?

**Comment by mordante** [comment]
CMake currently has experimental support for modules in their master branch. I've update our modules status page. Do you still want to pursue this patch? If not I'll close the patch.

**Comment by a858438680** [comment]
> CMake currently has experimental support for modules in their master branch. I've update our modules status page. Do you still want to pursue this patch? If not I'll close the patch.

Actually I looked the newest document. And I think this patch is not maintainable, giving that cmake now has their official method. So I don't want to pursue this patch. It's completely ok for me to close this.

**Comment by mordante** [comment]
> > CMake currently has experimental support for modules in their master branch. I've update our modules status page. Do you still want to pursue this patch? If not I'll close the patch.
> 
> Actually I looked the newest document. And I think this patch is not maintainable, giving that cmake now has their official method. So I don't want to pursue this patch. It's completely ok for me to close this.

Thanks then I'll close this.
 
> And Actually I tried to rebase this path to the main branch. But now I don't know how to do it.

You need to resolve the conflicts to be able to rebase. If you want to know more about this, let me know.
